shakenfist/daemons/cluster.py

```diff
@@ -53,21 +56,116 @@ def _await_election(self):
+                    LOG.with_fields({
+                        'blob': b,
+                        'ignored_locations': ignored_locations
+                    }).info('Ignored some blob locations as nodes are absent')
+
+                delta = len(locations) - config.BLOB_REPLICATION_FACTOR
+                if delta > 0:
+                    excess_locations = b.locations
+                    in_use_locations = []
+
+                    for instance_uuid in b.instances:
+                        i = instance.Instance.from_db(instance_uuid)
+                        node = i.placement.get('node')
+                        if node in excess_locations:
+                            excess_locations.remove(node)
+                            in_use_locations.append(node)
+
+                    if len(in_use_locations) >= config.BLOB_REPLICATION_FACTOR:
+                        # We have enough locations in use to meet the replication
+                        # factor, remove everything else
+                        overreplicated[b.uuid] = excess_locations
+                    else:
+                        # We need to keep some idle locations to maintain our
+                        # replication factor. Prefer to remove the fullest first.
+                        overreplicated[b.uuid] = []
+                        target = config.BLOB_REPLICATION_FACTOR - \
+                            len(in_use_locations)
+                        for n in nodes_by_disk_ascending:
+                            if n in excess_locations:
+                                overreplicated[b.uuid].append(n)
+                            if len(overreplicated[b.uuid]) == target:
+                                break
+
+                elif delta < 0:
+                    # The tuple is blob UUID, and how much to over replicate by.
+                    underreplicated.append((b.uuid, 0))
+
+                else:
+                    # We have exactly the right number of copies, but what if
+                    # the blob is on a really full node?
+                    for n in low_disk_nodes:
+                        if n in b.locations:
+                            # We have at least one space constrained node with
+                            # this blob. Request an extra temporary copy of the
+                            # blob elsewhere so we can hopefully clean up one of
+                            # these next pass. The tuple is blob UUID, and how
+                            # much to over replicate by.
+                            underreplicated.append((b.uuid, 1))
+                            break
+
+        for blob_uuid in overreplicated:
+            b = blob.Blob.from_db(blob_uuid)
+            for node in overreplicated[blob_uuid]:
+                LOG.with_fields({
+                    'blob': b,
+                    'node': node
+                }).info('Blob over replicated, removing from node with no users')
+                b.drop_node_location(node)
```

Comment:
So you missed the bit I thought would trigger you -- this block keeps blob replication counts down to as close to the specified number as possible. Now, it does remove from the fullest nodes first, but if you're drowning in free disk it still cleans up. You were arguing a couple of weeks ago that we should only cleanup if we're disk constrained, and otherwise over replication wasn't a problem. Discuss.

Reply:
I was thinking this was the drip-feed of a larger branch?

Yes, it would be good to specify an upper and lower disk space bound before moving blobs.

Reply:
It is the drip feed of a bigger series, but we can definitely change things along the way if we need to. That's the point of making it more reviewable.